scipy/optimize/optimize.py

```diff
@@ -2397,20 +2399,76 @@ def bracket(func, xa=0.0, xb=1.0, args=(), grow_limit=110.0, maxiter=1000):
     return xa, xb, xc, fa, fb, fc, funcalls
 
 
-def _linesearch_powell(func, p, xi, tol=1e-3):
+def _line_for_search(x0, alpha, lower_bound, upper_bound):
+    """
+    Given an n-dimensional parameter vector x0 and an n-dimensional
+    direction vector alpha, and lower and upper bounds on each of
+    the n parameters, what are the bounds on a scalar l such that
+    lower_bound <= x0 + alpha * l <= upper_bound.
+
+    x0 is the vector representing the current location (np.array)
+    alpha is the vector representing the direction (np.array)
+
+    ie the direction is along the line from x0 to alpha
+
+    lower_bound is a np.array of the lower bounds for each parameter in x0.
+    upper_bound is a np.array of the upper bounds for each parameter in x0.
+
+    returns (lmin, lmax), the bounds for l such that
+            lower_bound[i] <= x0_i+alpha_i * l <= upper_bound[i]
+        for all i.
+    """
+    # get nonzero indices of alpha so we don't get any zero division errors.
+    # alpha will not be all zero, since it is called from _linesearch_powell
+    # where we have a check for this.
+    nonzero, = alpha.nonzero()
+    lower_bound, upper_bound = lower_bound[nonzero], upper_bound[nonzero]
+    x0, alpha = x0[nonzero], alpha[nonzero]
+    low = (lower_bound - x0) / alpha
+    high = (upper_bound - x0) / alpha
+
+    # positive and negative indicies
+    pos, neg = alpha > 0, alpha < 0
+    lmin = np.max(low * pos + high * neg)
+    lmax = np.min(low * neg + high * pos)
+
+    return lmin, lmax
+
+
+def _linesearch_powell(func, p, xi, fval,
+                       lower_bound=None, upper_bound=None, tol=1e-3):
     """Line-search algorithm using fminbound.
 
     Find the minimium of the function ``func(x0+ alpha*direc)``.
+    
+    fval = func(p), might as well avoid recomputing it (limit fevals).
+
+    lower_bound and upper_bound should be np.arrays, where
+    lower_bound[i] is the lower bounds on the i^th parameter,
+    etc.
 
     """
     def myfunc(alpha):
         return func(p + alpha*xi)
-    alpha_min, fret, iter, num = brent(myfunc, full_output=1, tol=tol)
-    xi = alpha_min*xi
-    return squeeze(fret), p + xi, xi
 
+    # if xi is zero, then don't optimize
+    if not np.any(xi):
+        return fval, p, np.zeros(xi.shape)
+    elif lower_bound is None and upper_bound is None:
+        alpha_min, fret, iter, num = brent(myfunc, full_output=1, tol=tol)
+        xi = alpha_min*xi
+        return squeeze(fret), p + xi, xi
+    else:
+        bound = _line_for_search(p, xi, lower_bound, upper_bound)
+        res = _minimize_scalar_bounded(myfunc, bound, xatol=tol / 100)
+        xi = res.x * xi
+        if res.fun <= fval:
+            return squeeze(res.fun), p + xi, xi
```

Comment:
what's the squeeze here for?

Reply:
I included it because it had previously been included in the same function, the lines

```python
alpha_min, fret, iter, num = brent(myfunc, full_output=1, tol=tol)
xi = alpha_min*xi
return squeeze(fret), p + xi, xi
```
Without `squeeze(res.fun)`, minimizing a single parameter function with `r = minimize(f, x, method="Powell", ...)` where `x` is just a number (ie not `[x]`) will cause `r.fun` to be a 1 element array `array([number])` rather than just `number` or `array(number)`.